archive/100-theorems-list/90_stirlings_formula.lean

```diff
@@ -0,0 +1,244 @@
+  1 / 2 * log (2 * n.succ) - n.succ * log (n.succ / exp 1) :=
+begin
+  have h3, from sqrt_ne_zero'.mpr (mul_pos two_pos $ cast_pos.mpr (succ_pos n)),
+  have h4 : 0 ≠ ((n.succ : ℝ) / exp 1) ^ n.succ, from
+    ne_of_lt (pow_pos (div_pos (cast_pos.mpr n.succ_pos ) (exp_pos 1)) n.succ),
+  rw [log_stirling_seq, stirling_seq, log_div, log_mul, sqrt_eq_rpow, log_rpow, log_pow],
+  { linarith },
+  { exact (zero_lt_mul_left zero_lt_two).mpr (cast_lt.mpr n.succ_pos),},
+  { exact h3, },
+  { exact h4.symm, },
+  { exact cast_ne_zero.mpr n.succ.factorial_ne_zero, },
+  { apply (mul_ne_zero h3 h4.symm), },
+end
+
+/--
+The sequence `log_stirling_seq (m + 1) - log_stirling_seq (m + 2)` has the series expansion
+   `∑ 1 / (2 * (k + 1) + 1) * (1 / 2 * (m + 1) + 1)^(2 * (k + 1))`
+-/
+lemma log_stirling_seq_diff_has_sum (m : ℕ) :
+  has_sum (λ k : ℕ, (1 : ℝ) / (2 * k.succ + 1) * ((1 / (2 * m.succ + 1)) ^ 2) ^ k.succ)
+  (log_stirling_seq m.succ - log_stirling_seq m.succ.succ) :=
+begin
+  change
+    has_sum ((λ b : ℕ, 1 / (2 * (b : ℝ) + 1) * ((1 / (2 * m.succ + 1)) ^ 2) ^ b) ∘ succ) _,
+  rw has_sum_nat_add_iff 1,
+  convert (power_series_log_succ_div $ cast_pos.mpr (succ_pos m)).mul_left ((m.succ : ℝ) + 1 / 2),
+  { ext k,
+    rw [← pow_mul, pow_add],
+    have : 2 * (k : ℝ) + 1 ≠ 0, by {norm_cast, exact succ_ne_zero (2*k)},
+    have : 2 * (m.succ : ℝ) + 1 ≠ 0, by {norm_cast, exact succ_ne_zero (2*m.succ)},
+    field_simp,
+    ring },
+  { have h_reorder : ∀ {a b c d e f : ℝ}, a - 1 / 2 * b - c - (d - 1 / 2 * e - f) =
+      (a - d) - 1 / 2  * (b - e) - (c - f), by
+    { intros, ring_nf },
+    rw [log_stirling_seq_formula, log_stirling_seq_formula, h_reorder],
+    repeat {rw [log_div, factorial_succ]},
+    push_cast,
+    repeat {rw log_mul},
+    rw log_exp,
+    ring_nf,
+    all_goals {norm_cast},
+    { rw [range_one, sum_singleton, mul_zero, zero_add, cast_one, pow_zero, one_mul, inv_one,
+        sub_self, add_zero] },
+    all_goals {try {refine mul_ne_zero _ _}, try {exact succ_ne_zero _}},
+    any_goals {exact factorial_ne_zero m},
+    any_goals {exact exp_ne_zero 1} },
+  { apply_instance }
+end
+
+/-- The sequence `log_stirling_seq ∘ succ` is monotone decreasing -/
+lemma log_stirling_seq'_antitone : antitone (log_stirling_seq ∘ succ) :=
+begin
+  apply antitone_nat_of_succ_le,
+  intro n,
+  refine sub_nonneg.mp _,
+  rw ← succ_eq_add_one,
+  refine has_sum.nonneg _ (log_stirling_seq_diff_has_sum n),
+  norm_num,
+  simp only [one_div],
+  intro m,
+  refine mul_nonneg _ _,
+  all_goals {refine inv_nonneg.mpr _, norm_cast, exact (zero_le _)},
+end
+
+/--
+We have the bound  `log_stirling_seq n - log_stirling_seq (n+1) ≤ 1/(2n+1)^2* 1/(1-(1/2n+1)^2)`.
+-/
+lemma log_stirling_seq_diff_le_geo_sum (n : ℕ) :
+  log_stirling_seq n.succ - log_stirling_seq n.succ.succ ≤
+  (1 / (2 * n.succ + 1)) ^ 2 / (1 - (1 / (2 * n.succ + 1)) ^ 2) :=
+begin
+  have h_nonneg : 0 ≤ ((1 / (2 * (n.succ : ℝ) + 1)) ^ 2), by
```

Comment:
Not sure we have a fixed rule here, but since the `by` in `have ..., by { ... }` is unnecessary, most people omit it.  (Same in several places below.)

Reply:
Good to know, removed all occurrences. 
Also I removed `by` in calc mode if there are curly braces afterwards. Not sure if that is also more idiomatic....

Reply:
(Doesn't work in calc mode, reverting...)